spacy/training/align.py

```diff
@@ -32,3 +30,78 @@ def _make_ragged(indices):
     for x in indices:
         flat.extend(x)
     return Ragged(numpy.array(flat, dtype="i"), lengths)
+
+
+def get_alignments(A: List[str], B: List[str]) -> Tuple[List[List[int]], List[List[int]]]:
+    """Calculate alignment tables between two tokenizations.
+
+    tokens_a (List[str]): The first tokenization.
+    tokens_b (List[str]): The second tokenization.
+    RETURNS: (tuple): A tuple consisting of the following information:
+      * a2b (List[List[int]): Mapping of indices in `tokens_a` to indices in
+        `tokens_b`. For instance, if `a2b[4] == [6]`, that means that
+        `tokens_a[4]` aligns to `tokens_b[6]`.
+      * b2a (List[List[int]]): The same as `a2b`, but mapping the other
+        direction.
+    """
+    # Remove whitespace from all non-whitespace tokens, replace all whitespace
+    # tokens with a single space
+    tokens_a = [x.replace(" ", "") if not x.isspace() else " " for x in A]
+    tokens_b = [x.replace(" ", "") if not x.isspace() else " " for x in B]
+    # Handle leading whitespace: replace whitespace tokens with empty strings
+    # unless the whitespace is in both A and B
+    str_a = "".join(tokens_a)
+    str_b = "".join(tokens_b)
+    _adjust_whitespace(tokens_a, str_a, str_b)
+    _adjust_whitespace(tokens_b, str_b, str_a)
+    # Handle trailing whitespace: temporarily split off trailing whitespace
+    # tokens and align them in the same way as leading whitespace
+    trailing_ws_start_a = _get_trailing_ws_start(tokens_a)
+    trailing_ws_start_b = _get_trailing_ws_start(tokens_b)
+    trailing_ws_a = tokens_a[trailing_ws_start_a:]
+    trailing_ws_b = tokens_b[trailing_ws_start_b:]
+    str_trailing_ws_a = "".join(trailing_ws_a)
+    str_trailing_ws_b = "".join(trailing_ws_b)
+    _adjust_whitespace(trailing_ws_a, str_trailing_ws_a, str_trailing_ws_b)
+    _adjust_whitespace(trailing_ws_b, str_trailing_ws_b, str_trailing_ws_a)
+    tokens_a = tokens_a[:trailing_ws_start_a] + trailing_ws_a
+    tokens_b = tokens_b[:trailing_ws_start_b] + trailing_ws_b
+    # Check that the strings align, allowing only differences in capitalization
+    if "".join(tokens_a).lower() != "".join(tokens_b).lower():
+        raise ValueError(Errors.E949)
+    # Store the set of aligned token indices per token
+    a2b = [set() for _ in range(len(tokens_a))]
+    b2a = [set() for _ in range(len(tokens_b))]
+    # Create iterators for token indices per character position
+    token_by_char_a = chain(*[[i] * len(x) for i, x in enumerate(tokens_a)])
```

Comment:
Is this fast enough? It looks terrible.

Reply:
It does look terrible. I'll refactor it a bit.